### PR TITLE
P2-2106-AbTesting_job_api

### PR DESCRIPTION
**Pull request by karen-hanna**

### CODE REVIEW

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.ApprenticeFeedback.Application/Commands/ProcessFeedbackTargetVariants/ProcessFeedbackTargetVariantsCommandHandler.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            if(request.MergeStaging) 
+            {
+                await MergeFeedbackTargetVariantsAsync();
+            }
+
+            return Unit.Value;
+
+        }
+
+        private async Task ClearStagingData()
```
[the remark this answers is not in the document]
Comment (karen-hanna) [reply src/SFA.DAS.ApprenticeFeedback.Application/Commands/ProcessFeedbackTargetVariants/ProcessFeedbackTargetVariantsCommandHandler.cs:51]: Added stored procedures for truncate and merge

File: src/SFA.DAS.ApprenticeFeedback.Api.IntegrationTests/Handlers/FeedbackTargetVariantStagingHandler.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        public static async Task<FeedbackTargetVariantModel> QueryFirstOrDefaultAsync(FeedbackTargetVariantModel feedbackTransaction)
+        {
+            var sqlToQuery =
+                "SELECT " +
+                    "[ApprenticeshipId]" +
+                    ", [Variant]" +
+                "FROM [FeedbackTargetVariant_Staging] " +
+                //$"WHERE (Id = @id OR @id IS NULL) " + // when @id is null then Id is not predicated
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.ApprenticeFeedback.Api.IntegrationTests/Handlers/FeedbackTargetVariantStagingHandler.cs:34]: Commented out code

File: src/SFA.DAS.ApprenticeFeedback.Api.IntegrationTests/Handlers/FeedbackTargetVariantStagingHandler.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                    $"AND {NotNullQueryParam(feedbackTransaction, p => p.ApprenticeshipId)} " +
+                    $"AND {NullQueryParam(feedbackTransaction, p => p.Variant)} ";
+
+            return await DatabaseService.QueryFirstOrDefaultAsync<FeedbackTargetVariantModel, FeedbackTargetVariantModel>(sqlToQuery, feedbackTransaction);
+        }
+
+        public static string Because(FeedbackTargetVariantModel feedbackTransaction)
+        {
+            var becauseMessage = $"Expected a FeedbackTransaction to exist with parameters: " +
+                                 //$"{BecauseParam(feedbackTransaction, p => p.Id, allowAny: true)}, " +
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.ApprenticeFeedback.Api.IntegrationTests/Handlers/FeedbackTargetVariantStagingHandler.cs:44]: And here

File: src/SFA.DAS.ApprenticeFeedback.Database/StoredProcedures/ImportIntoFeedbackTargetVariant_FromFeedbackTargetVariant_Staging.sql [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    MERGE INTO [dbo].[FeedbackTargetVariant] AS TARGET
+    USING [dbo].[FeedbackTargetVariant_Staging] AS SOURCE
+    ON TARGET.ApprenticeshipId = SOURCE.ApprenticeshipId
+    WHEN NOT MATCHED BY TARGET THEN
+        INSERT (ApprenticeshipId, Variant)
+        VALUES (SOURCE.ApprenticeshipId, SOURCE.Variant)
+    WHEN MATCHED THEN
+        UPDATE SET TARGET.Variant = SOURCE.Variant;
+
+    DELETE FROM [dbo].[FeedbackTargetVariant]
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.ApprenticeFeedback.Database/StoredProcedures/ImportIntoFeedbackTargetVariant_FromFeedbackTargetVariant_Staging.sql:13]: WHEN NOT MATCHED BY SOURCE THEN
        DELETE; 

You can do the delete as part of the merge

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.ApprenticeFeedback.Api.IntegrationTests/Handlers/FeedbackTargetVariantStagingHandler.cs (hunk 2 above)
Comment (karen-hanna) [reply src/SFA.DAS.ApprenticeFeedback.Api.IntegrationTests/Handlers/FeedbackTargetVariantStagingHandler.cs:34]: removed

File: src/SFA.DAS.ApprenticeFeedback.Database/StoredProcedures/ImportIntoFeedbackTargetVariant_FromFeedbackTargetVariant_Staging.sql (hunk 4 above)
Comment (karen-hanna) [reply src/SFA.DAS.ApprenticeFeedback.Database/StoredProcedures/ImportIntoFeedbackTargetVariant_FromFeedbackTargetVariant_Staging.sql:13]: updated

**Review by ChrisJWoodcock — APPROVED** [review]

**Pull request merged**